model/riscv_insts_vext_utils.sail

```diff
@@ -0,0 +1,703 @@
+  }
+  else {
+    elem_sat = slice(elem, 0, 'm);
+    writeCSR(csr_name_map("vxsat"), EXTZ(0b0))
+  };
+  elem_sat
+}
+
+/* Get the floating point rounding mode from csr fcsr */
+val get_fp_rounding_mode : unit -> rounding_mode effect {rreg}
+function get_fp_rounding_mode() = encdec_rounding_mode(fcsr.FRM())
+
+/* Split sign and the remain of floating point number */
+val      fsplitsign : forall 'n, 'n in {16, 32, 64}. bits('n) -> (bits(1), bits('n - 1))
+function fsplitsign   (xf) = {
+  match 'n{
+    16 => (xf[15..15], xf[14..0]),
+    32 => (xf[31..31], xf[30..0]),
+    64 => (xf[63..63], xf[62..0])
+  }
+}
+
+/* Make a floating point number by sign and the remains bits */
+val      fmakesign : forall 'n, 'n in {16, 32, 64}. (bits(1), bits('n - 1)) -> bits('n)
+function fmakesign (sign, remain) = sign @ remain
+
+/* Negate a floating point number */
+val negate_fp : forall 'n, 'n in {16, 32, 64}. bits('n) -> bits('n) 
+function negate_fp (xf) = {
+  let (sign, remain) = fsplitsign(xf);
+  let new_sign = if (sign == 0b0) then 0b1 else 0b0;
+  fmakesign (new_sign, remain)
+}
+
+
+/* Floating point functions */
+val fp_add: forall 'n, 'n in {16, 32, 64}. (bits(3), bits('n), bits('n)) -> bits('n) effect {escape, rreg, undef, wreg}
+function fp_add(rm_3b, op1, op2) = {
+  fflags : bits_fflags = undefined;
+  result_val : bits('n) = undefined;
+  (fflags, result_val) = match 'n {
+    16  => riscv_f16Add(rm_3b, op1, op2),
+    32  => riscv_f32Add(rm_3b, op1, op2),
+    64  => riscv_f64Add(rm_3b, op1, op2)
+  };
+  write_fflags(fflags);
+  result_val
+}
+
+val fp_sub: forall 'n, 'n in {16, 32, 64}. (bits(3), bits('n), bits('n)) -> bits('n) effect {escape, rreg, undef, wreg}
+function fp_sub(rm_3b, op1, op2) = {
+  fflags : bits_fflags = undefined;
+  result_val : bits('n) = undefined;
+  (fflags, result_val) = match 'n {
+    16  => riscv_f16Sub(rm_3b, op1, op2),
+    32  => riscv_f32Sub(rm_3b, op1, op2),
+    64  => riscv_f64Sub(rm_3b, op1, op2)
+  };
+  write_fflags(fflags);
+  result_val
+}
+
+val fp_min : forall 'n, 'n in {16, 32, 64}. (bits('n), bits('n)) -> bits('n) effect {escape, rreg, undef, wreg}
+function fp_min(op1, op2) = {
+  temp_flags : bits_fflags = undefined;
+  op1_lt_op2 : bool = undefined;
+
+  (temp_flags, op1_lt_op2) = match 'n {
+    16  => riscv_f16Lt(op1, op2),
+    32  => riscv_f32Lt(op1, op2),
+    64  => riscv_f64Lt(op1, op2)
+  };
+
+  let result_val  = if (f_is_NaN(op1) & f_is_NaN(op2)) then canonical_NaN('n)
+                else if f_is_NaN(op1) then op2
+                else if f_is_NaN(op2) then op1
+                else if (f_is_neg_zero(op1) & f_is_pos_zero(op2)) then op1
+                else if (f_is_neg_zero(op2) & f_is_pos_zero(op1)) then op2
+                else if op1_lt_op2 then op1
+                else op2;
+
+  let fflags = if (f_is_SNaN(op1) | f_is_SNaN(op2)) then nvFlag() else zeros();
+  accrue_fflags(fflags);
+  update_softfloat_fflags(fflags); /* to avoid inconsistency about NaN operands */
```

Comment:
accrue_fflags already does that?

Reply:
The implementation of `vfmax` and `vfmin` currently use the softfloat code of `flt` functions but they have different judgements for invalid operation regarding NaN operands, so I have to temporarily add this line. But I have found that there are `flt_quiet` functions in the softfloat code that can perfectly fit `vfmax` and `vfmin`. I will update the `fp_max` and `fp_min` functions here in the floating-point instruction PR later. They don't have to be written like this.

Reply:
If you're abusing soft float interfaces intended for executing architectural floating-point instructions in implementation-specific parts of the model and clobbering architectural parts of the model as a result then that needs fixing